### PR TITLE
Adds support for SQL Server specific types and syntax

### PR DESCRIPTION
**Pull request by wwwiiilll**
SQL Server uses different syntax/types than others, this includes fixes and improvements for:

- fixed creating tables with unbounded `varchar` or `char` columns
- fixed creating tables with a `double` column
- fixed incorrect usage of `timestamp` (aka `rowversion`) instead of `datetime2` for Presto `timestamp` columns
- fixed inserting `real` columns
- fixed inserting `timestamp` and `time` columns
- added `datetimeoffset` mapping for `timestamp with time zone`
- added `bit` mapping for `boolean`

Also changes the documentation to adds all the supported versions and the full data type mapping table (there are more, `rowversion` for instance, but they don't have viable representations in Java and are therefore not really useful outside T-SQL).

Unfortunately it's mostly duplication of the `base-jdbc` since I needed access to the original ResultSet and PreparedStatement to call some non-standard methods on them. This was actually originally an internal plugin here so I had to work with what was available, but I could change `base-jdbc` to raise the visibility of what I need if it's desirable.

### CODE REVIEW

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java [hunk 1]
```diff
@@ -18,49 +18,148 @@
...
+            return sqlType;
+        }
+        throw new PrestoException(NOT_SUPPORTED, "Unsupported column type: " + type.getTypeSignature());
+    }
+
+    @Override
+    protected String quoted(String name)
     {
-        return "\'" + literal + "\'";
+        return '[' + name.replace("]", "]]") + ']';
```
Comment (findepi) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java]: I'd like to know whether there really are cases when `"`-quoting doesn't work and we have to speak `[]`-slang.

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerConnectorFactory.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    @Override
+    public ConnectorHandleResolver getHandleResolver()
+    {
+        return new JdbcHandleResolver();
+    }
+
+    @Override
+    public Connector create(String connectorId, Map<String, String> requiredConfig, ConnectorContext context)
+    {
+        requireNonNull(requiredConfig, "requiredConfig is null");
```
Comment (findepi) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerConnectorFactory.java]: check other args are not null too?

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerPlugin.java [hunk 3]
```diff
@@ -13,13 +13,23 @@
...
-import com.facebook.presto.plugin.jdbc.JdbcPlugin;
+import com.facebook.presto.spi.Plugin;
+import com.facebook.presto.spi.connector.ConnectorFactory;
+import com.google.common.collect.ImmutableList;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
 
 public class SqlServerPlugin
-        extends JdbcPlugin
+        implements Plugin
```
Comment (findepi) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerPlugin.java]: * All changes related to metadata, binders plugin scaffolding in the sql server plugin should go into separate commit.
* also, why no longer `extends JdbcPlugin`?

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerRecordCursor.java [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    }
+
+    @Override
+    public Type getType(int field)
+    {
+        return this.columnHandles.get(field).getColumnType();
+    }
+
+    @Override
+    public boolean advanceNextPosition()
```
Comment (findepi) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerRecordCursor.java]: The body of this method looks very much like `com.facebook.presto.plugin.jdbc.JdbcRecordCursor#advanceNextPosition`. Perhaps there is a possibility to retain reuse?

Same applies to other methods too.

Even if you need to copy something (i hope you don't), clearly separate copying and enhancements into two separate commits, otherwise I don't know what to review.

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerRecordSet.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import com.facebook.presto.plugin.jdbc.JdbcColumnHandle;
+import com.facebook.presto.plugin.jdbc.JdbcRecordSet;
+import com.facebook.presto.plugin.jdbc.JdbcSplit;
+import com.facebook.presto.spi.RecordCursor;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+class SqlServerRecordSet
```
Comment (findepi) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerRecordSet.java]: Introduction of this class & plumbing around it should also be separate commit from data type changes.

File: presto-docs/src/main/sphinx/connector/sqlserver.rst [hunk 6]
```diff
@@ -67,22 +67,39 @@ that catalog name instead of ``sqlserver`` in the above examples.
...
+``smallint``                  ``smallint``
+``tinyint``                   ``tinyint``
+``real``                      ``real``
 ``float``                     ``double``
+``decimal``                   ``double``
+``numeric``                   ``double``
 ``char(n)``                   ``char(n)``
+``nchar(n)``                  ``char(n)``
 ``varchar(n)``                ``varchar(n)``
+``nvarchar(n)``               ``varchar(n)``
```
Comment (wwwiiilll) [inline presto-docs/src/main/sphinx/connector/sqlserver.rst]: That's the thing, if you specify `varchar` without length in T-SQL it's equivalent to `varchar(8000)` in Presto whilst `varchar(max)` is the unbounded (2GB) `varchar` in Presto, same goes for `char`. Maybe I should indeed add the `varchar(max)` = `varchar` lines.

File: presto-docs/src/main/sphinx/connector/sqlserver.rst [hunk 7]
```diff
@@ -67,22 +67,39 @@ that catalog name instead of ``sqlserver`` in the above examples.
...
+``text``                      ``varchar``
+``ntext``                     ``varchar``
 ``date``                      ``date``
+``time``                      ``time``
+``datetime``                  ``timestamp``
+``smalldatetime``             ``timestamp``
+``datetime2``                 ``timestamp``
+``datetimeoffset``            ``timestamp with time zone``
+``money``                     ``double``
```
Comment (wwwiiilll) [inline presto-docs/src/main/sphinx/connector/sqlserver.rst]: It's the type returned by the JDBC driver, I didn't do anything special with it actually.

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java [hunk 8]
```diff
@@ -18,49 +18,148 @@
...
         }
     }
 
-    private static String singleQuote(String catalog, String schema, String table)
+    @Override
+    public PreparedStatement getPreparedStatement(Connection connection, String sql)
+            throws SQLException
     {
-        return singleQuote(catalog + "." + schema + "." + table);
+        connection.setAutoCommit(false);
```
Comment (wwwiiilll) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java]: It's enabled by default, since we'll manage the transaction in any case or we won't need it, it removes _some_ very minimal overhead.

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java [hunk 9]
```diff
@@ -18,49 +18,148 @@
...
+        PreparedStatement statement = connection.prepareStatement(sql);
+        statement.setFetchSize(1000);
```
Comment (wwwiiilll) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java]: It's from the PostgreSQL plugin, in this case it'll default to 128 which is a bit small for large data sets.

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java [hunk 10]
```diff
@@ -18,49 +18,148 @@
...
+        return statement;
     }
 
-    private static String singleQuote(String literal)
+    @Override
+    protected Type toPrestoType(int jdbcType, int columnSize)
+    {
+        if (jdbcType == Types.DATETIMEOFFSET) {
```
Comment (wwwiiilll) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java]: `SQL_TYPES` does the opposite, Presto -> SQL Server

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java [hunk 11]
```diff
@@ -18,49 +18,148 @@
...
+    }
+
+    @Override
+    protected String toSqlType(Type type)
+    {
+        if (type instanceof VarcharType) {
+            if (((VarcharType) type).isUnbounded()) {
+                return "varchar(max)";
+            }
+            return "varchar(" + ((VarcharType) type).getLengthSafe() + ')';
```
Comment (wwwiiilll) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java]: My assumptions were wrong, went back to read the doc, >8000 has to map to `varchar(max)`, I'll fix that.

File: presto-docs/src/main/sphinx/connector/sqlserver.rst (hunk 6 above)
Comment (wwwiiilll) [inline presto-docs/src/main/sphinx/connector/sqlserver.rst]: I misread the docs, `varchar` (MSSQL) is actually `varchar(1)` (Presto), same for `char`, I'll update accordingly.

**Comment by findepi** [comment]
one more thing -- commit message - use imperative form (unfortunately https://github.com/prestodb/presto/#developers doesn't contain the guide...)

>> Unfortunately it's mostly duplication of the base-jdbc since I needed access to the original ResultSet and PreparedStatement to call some non-standard methods on them

> This is indeed unfortunate. @electrum would it be possible to guide this PR or change base JDBC connector (if needed) so that there is no need for massive copy&paste?

regarding this,  it doesn't stand out  what/where you actually needed, but this will become clear once you separate commits

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java (hunk 1 above)
Comment (wwwiiilll) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java]: I could actually leave that out, since `QUOTED_IDENTIFIER` is set to `ON`, `"`-quoting would work just fine.

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerPlugin.java (hunk 3 above)
Comment (wwwiiilll) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerPlugin.java]: I'm overriding everything in it, there's no need for this to be a `JdbcPlugin`.

**Comment by wwwiiilll** [comment]
I'll try to restructure this. Should I start out with a "base commit" with the copied stuff from `base-jdbc` and work my way from there?

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerRecordCursor.java (hunk 4 above)
Comment (wwwiiilll) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerRecordCursor.java]: Well it is copied indeed, without changing the visibility of the JdbcRecordCursor members I can't call  it's constructor, so I can inherit from it.

**Comment by findepi** [comment]
> I'll try to restructure this. Should I start out with a "base commit" with the copied stuff from base-jdbc and work my way from there? It's actually about 90% duplication.

That's good idea. Generally you have at hand the ± expected end state and in your workspace you apply refactors/changes towards that, committing at logical boundaries.

However, the duplication is worrisome and I think we should work on avoiding/reducing it. For example the base jdbc connector could take a data type from `ResultSet` extractor and you would override/extend only that. (Just thinking..) 
It's not my call, however, to decide how exactly this should be tackled.

**Comment by wwwiiilll** [comment]
Setting a couple of fields to `protected` in `base-jdbc` would effectively remove most if not all the duplication. If this is what we want to do, I'll gladly change it. Otherwise I'm open to any suggestion.

Apart from that, how is the structure right now? I tried to isolate actual changes from the duplication.

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java [hunk 12]
```diff
@@ -35,16 +35,19 @@ public SqlServerClient(JdbcConnectorId connectorId, BaseJdbcConfig config)
...
+        connectionProperties.setProperty("applicationName", "Presto");
     }
 
     @Override
     public void commitCreateTable(JdbcOutputTableHandle handle)
     {
         StringBuilder sql = new StringBuilder()
                 .append("sp_rename ")
-                .append(singleQuote(handle.getCatalogName(), handle.getSchemaName(), handle.getTemporaryTableName()))
+                .append("'" + quoted(handle.getCatalogName(), handle.getSchemaName(), handle.getTemporaryTableName()) + "'")
```
Comment (findepi) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java]: Didn;'t run it, but having both `"'" +` and `quoted()` looks suspicious

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java [hunk 13]
```diff
@@ -35,16 +35,19 @@ public SqlServerClient(JdbcConnectorId connectorId, BaseJdbcConfig config)
...
                 .append(", ")
-                .append(singleQuote(handle.getTableName()));
+                .append("'" + handle.getTableName() + "'");
```
Comment (findepi) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java]: Didn't run it, but above you had `quoted` here you don't have. 

The convenience function `singleQuote` looked useful in this case.

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java (hunk 1 above)
Comment (findepi) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java]: If you can leave it  out, leave it out :)

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java (hunk 12 above)
Comment (wwwiiilll) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java]: Yeah well that's a weird one with SQL Server, then end result has to look something like this:
```
EXEC sp_rename '[northwind].[dbo].[some [weird]] table / name]', 'new_name'
```
You have to quote the fully qualified table name and then pass that as a string literal. It also appears to work with `"`-quoting, but I haven't done extensive testing with it.

File: presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerRecordCursor.java (hunk 4 above)
Comment (electrum) [inline presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerRecordCursor.java]: The constructor is public. Are you talking about `handleSqlException`? It's fine to make that method protected. We'd rather not duplicate a lot of code just to avoid changing visibility.

**Comment by electrum** [comment]
I have a mostly complete change to allow the type handling for read and write to be pluggable. This is needed to properly support JSON for PostgreSQL. I'll see how close that is to being ready.

Is there anything else needed here? We'd like to avoid having to fork the entire connector.

**Comment by wwwiiilll** [comment]
I really just need to be able to call some non-standard stuff on the `PreparedStatement` for `RecordSink` and on the `ResultSet` for `RecordCursor`.
Is there a branch I can look at with those changes?
I'm pretty sure it'll enable what I need to do here.
